### PR TITLE
Refactored Graphdb into an interface

### PR DESCRIPTION
**Pull request by nathannaveen**
- Refactored `Client` in `pkg/assembler/graphdb/graphdb.go` into an interface so that is easier to test.

Signed-off-by: nathannaveen <42319948+nathannaveen@users.noreply.github.com>

### CODE REVIEW

**Review by mihaimaruseac — APPROVED** [review]
Thanks. This LGTMs.

Note that we're likely going to change this in the GraphQL refactor, see #326

**Comment by nathannaveen** [comment]
> I have a question about this - what is the possible other implementations of the interface - since the interface seems to be specific to neo4j as well. Is it possible to mock a neo4j.Session?

I was thinking that it doesn’t have to be specific to neo4j. Also `neo4j.Session` is an interface which can be mocked for testing. That was my idea behind this PR, if I haven’t understood something please let me know.

**Comment by lumjjb** [comment]
gotcha - I was thinking of this as an interface for general graphDB, but this makes sense in the context for your intent for `neo4j`, if we can rename it to `Neo4jClient` instead, if you don't mind helping to do some of the renaming to follow as well. Thanks!

**Comment by nathannaveen** [comment]
> if you don't mind helping to do some of the renaming to follow as well

@lumjjb, I don’t understand what you mean by the above, could you explain? Thanks!

**Comment by lumjjb** [comment]
Ah yes... so what i was thinking is that there is a bit of a lack of definition around the "Client" as a generic client. This is more so that the current code is structured in a way that it doesn't split up the individual graph db implementations (YET). So I was just thinking in the future, we could do a refactor like the following which would make it match more of a pluggable structure: 

1. Rename `Client` to `Neo4jClient`
2. Move go files in `pkg/assembler/graphdb/` (since it mainly contains just neo4j stuff) to a new package `pkg/assembler/graphdb/neo4j`
3. Move `pkg/assembler/graphdb.go` into its own package `pkg/assembler/graphdb/graphdb.go` (if its not too messy to pull apart)

Again, this is me asking for extra toppings, it just came up and wondering if we could take this opportunity to do this too, but if not, feel free to say it's not something you want to handle in this PR and we can defer to an issue.

**Comment by nathannaveen** [comment]
> Ah yes... so what i was thinking is that there is a bit of a lack of definition around the "Client" as a generic client. This is more so that the current code is structured in a way that it doesn't split up the individual graph db implementations (YET). So I was just thinking in the future, we could do a refactor like the following which would make it match more of a pluggable structure:
> 
> 1. Rename `Client` to `Neo4jClient`
> 2. Move go files in `pkg/assembler/graphdb/` (since it mainly contains just neo4j stuff) to a new package `pkg/assembler/graphdb/neo4j`
> 3. Move `pkg/assembler/graphdb.go` into its own package `pkg/assembler/graphdb/graphdb.go` (if its not too messy to pull apart)
> 
> Again, this is me asking for extra toppings, it just came up and wondering if we could take this opportunity to do this too, but if not, feel free to say it's not something you want to handle in this PR and we can defer to an issue.

Thanks for the explanation! I have done all the refactoring!

**Review by mihaimaruseac — APPROVED** [review]

File: cmd/guacone/cmd/certifier.go
Comment (pxp928) [inline cmd/guacone/cmd/certifier.go:55]: Looks like the tests are failing for a similar change on cmd/pubsub_test

File: pkg/assembler/graphdb/neo4j/graphdb.go
Comment (pxp928) [inline pkg/assembler/graphdb/neo4j/graphdb.go:38]: Maybe its my lack of understanding on this PR but shouldn't the interface be defined at like graphdb.go and neo4J be an implementation of that interface? So that we can have a pluggable nature with more graphDB integration in the future?

Comment (nathannaveen) [reply pkg/assembler/graphdb/neo4j/graphdb.go:38]: I agree, but which `graphdb.go` are you referring to? Are you referring to the one in the `neo4j` package, or the one in the `graphdb` package?

**Comment by mihaimaruseac** [comment]
I'd say don't focus too much on this as it is going to be fully refactored with the GraphQL interface

**Comment by lumjjb** [comment]
We just had a maintainers workshop to round up the details of the graphql
api. We're cleaning up the notes and discussions and going to open it up
for comment and then start adding the graphql code!

**Comment by naveensrinivasan** [comment]
> We just had a maintainers workshop
Is this only for maintainers? Can it not be open for everyone?

**Comment by mlieberman85** [comment]
This particular workshop was an in person workshop planned months in advance among the maintainers. We do have community meetings that everyone is welcome to attend.

**Comment by naveensrinivasan** [comment]
> This particular workshop was an in person workshop planned months in advance among the maintainers. We do have community meetings that everyone is welcome to attend.



> This particular workshop was an in person workshop planned months in advance among the maintainers. We do have community meetings that everyone is welcome to attend.

👍 Thanks

**Comment by pxp928** [comment]
Closing due to graphQL work that will replace this.

**Pull request closed without merging**
